File: buildSrc/src/main/java/wolkenschloss/download/DownloadTasks.java
package wolkenschloss.download;

import org.gradle.api.tasks.TaskContainer;

public class DownloadTasks {
    public static final String DOWNLOAD_DISTRIBUTION_TASK_NAME = "download";
    private static final String GROUP_NAME = "download";

    private final TaskContainer tasks;

    public DownloadTasks(TaskContainer tasks) {
        this.tasks = tasks;
    }

    public void register(BaseImageExtension baseImage) {
        tasks.register(
                DOWNLOAD_DISTRIBUTION_TASK_NAME,
                DownloadDistribution.class,
                task -> {
                    task.setGroup(GROUP_NAME);
                    task.setDescription("Downloads the base image to a cache for later use.");
                    task.getBaseImageLocation().convention(baseImage.getUrl());
                    task.getDistributionName().convention(baseImage.getName());
                    task.getBaseImage().convention(baseImage.getBaseImageFile());
                    task.getDistributionDir().convention(baseImage.getDistributionDir());
                    task.getOutputs().upToDateWhen(spec -> task.getDistributionDir().get().getAsFile().exists());
                });
    }
}

File: buildSrc/src/main/kotlin/wolkenschloss/gradle/testbed/download/DownloadTasks.kt
package wolkenschloss.gradle.testbed.download

import org.gradle.api.tasks.TaskContainer

class DownloadTasks(private val extension: BaseImageExtension) {
    fun register(tasks: TaskContainer) {
        tasks.register(DOWNLOAD_DISTRIBUTION_TASK_NAME, DownloadDistribution::class.java) {
            group = GROUP_NAME
            description = "Downloads the base image to a cache for later use."
            baseImageLocation.convention(extension.url)
            distributionName.convention(extension.name)
            baseImage.convention(extension.baseImageFile)
            distributionDir.convention(extension.distributionDir)
            outputs.upToDateWhen { distributionDir.get().asFile.exists() }
        }
    }

    companion object {
        const val DOWNLOAD_DISTRIBUTION_TASK_NAME = "download"
        private const val GROUP_NAME = "download"
    }
}